### PR TITLE
Use the dlang.org's XML stream at feedburner to convert the latest articles to ddoc macros

### PR DESCRIPTION
**Pull request by wilzbach**
This won't work on DAutoTest, but the generated Ddoc file will be e.g.

```
DBLOG_LATEST_TITLE_1=DConf 2018 Programme &amp; Open Registration
DBLOG_LATEST_LINK_1=https://dlang.org/blog/2018/03/19/dconf-2018-programme-open-registration/
DBLOG_LATEST_DATE_1=March 19, 2018
DBLOG_LATEST_AUTHOR_1=Michael Parker
DBLOG_LATEST_TITLE_2=User Stories: Funkwerk
DBLOG_LATEST_LINK_2=https://dlang.org/blog/2018/03/14/user-stories-funkwerk/
DBLOG_LATEST_DATE_2=March 14, 2018
DBLOG_LATEST_AUTHOR_2=Michael Parker
```

### CODE REVIEW

File: posix.mak [hunk 1]
```diff
@@ -592,21 +592,12 @@ $G/dlangspec.verbatim.txt : $G/dlangspec-consolidated.d $(DDOC_BIN) verbatim.ddo
...
-	cat $< | grep -m2 'entry-title' | tail -n1 | \
-		sed -E 's/^.*<a href="(.+)" rel="bookmark">([^<]+)<\/a>.*<time.*datetime="[^"]+">([^<]*)<\/time><time class="updated".*Author *<\/span><a [^>]+>([^<]+)<\/a>.*/DBLOG_LATEST_TITLE2=\2|DBLOG_LATEST_LINK2=\1|DBLOG_LATEST_DATE2=\3|DBLOG_LATEST_AUTHOR2=\4/' | \
-		tr '|' '\n' > $(basename $@)2.ddoc
-	@if [ ! -s $(basename $@)2.ddoc ] ; then echo "$(basename $@)2.ddoc is empty. Please check the download"; rm -f $(basename $@)2.ddoc; exit 1 ; fi
-
-$G/dblog_latest2.ddoc: $G/dblog_latest.ddoc
+	curl -s --fail --retry 3 --retry-delay 5 -L http://feeds.feedburner.com/OfficialDBlog -o $@
+
+$G/dblog_latest.ddoc: $G/dblog_latest.xml $(STABLE_DMD) tools/ddoc_xml_extractor.d
+	$(STABLE_RDMD) tools/ddoc_xml_extractor.d -i $< -o $@
```
Comment (wilzbach) [inline posix.mak]: I used `--output` to avoid creating the file if an error happened

**Comment by CyberShadow** [comment]
> This won't work on DAutoTest, but the generated Ddoc file will be e.g.

One thing we could do is in DIFFABLE mode, use the same code path, but process a dummy static file. That'll allow testing the conversion on CI.

File: tools/ddoc_xml_extractor.d [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+auto toDdoc(const ref Article article, string suffix)
+{
+    string text;
+with(article)
+    text =
+`DBLOG_LATEST_TITLE%5$s=%1$s
+DBLOG_LATEST_LINK%5$s=%2$s
+DBLOG_LATEST_DATE%5$s=%3$s
+DBLOG_LATEST_AUTHOR%5$s=%4$s`.format(title, link, date, author, suffix);
```
Comment (CyberShadow) [inline tools/ddoc_xml_extractor.d]: Any of these need HTML/XML escaping?

Comment (wilzbach) [reply tools/ddoc_xml_extractor.d]: It looks to me like they are already escaped:

```
<title>DConf 2018 Programme &amp; Open Registration</title>
<link>https://dlang.org/blog/2018/03/19/dconf-2018-programme-open-registration/</link>
<comments>https://dlang.org/blog/2018/03/19/dconf-2018-programme-open-registration/#respond</comments>
<pubDate>Mon, 19 Mar 2018 13:27:32 +0000</pubDate>
<dc:creator><![CDATA[Michael Parker]]></dc:creator>
```

**Comment by wilzbach** [comment]
> One thing we could do is in DIFFABLE mode, use the same code path, but process a dummy static file. That'll allow testing the conversion on CI.

Done.

File: tools/ddoc_xml_extractor.d (hunk 2 above)
Comment (CyberShadow) [reply tools/ddoc_xml_extractor.d]: > `<title>DConf 2018 Programme &amp; Open Registration</title>`

Won't std.xml parse that `&amp;` into an `&`, meaning we now have to escape it back into an `&amp;` ?

File: dblog_feed_example.xml [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		<pubDate>Mon, 12 Mar 2018 14:20:54 +0000</pubDate>
+		<dc:creator><![CDATA[Michael Parker]]></dc:creator>
+				<category><![CDATA[Community]]></category>
+		<category><![CDATA[D Foundation]]></category>
+
+		<guid isPermaLink="false">http://dlang.org/blog/?p=1441</guid>
+		<description><![CDATA[In its work guiding the development of D and promoting its adoption, the D Language Foundation is driven primarily by donations big and small. The money comes in from different sources, the most visible being those listed on the website’s donation page, and is put to use in different ways. Today, the D Language Foundation is opening a new chapter in the donation story.]]></description>
+		<wfw:commentRss>https://dlang.org/blog/2018/03/12/the-d-foundation-at-open-collective/feed/</wfw:commentRss>
+		<slash:comments>0</slash:comments>
+	<post-id xmlns="com-wordpress:feed-additions:1">1441</post-id>	</item>
```
Comment (CyberShadow) [inline dblog_feed_example.xml]: Cool, thanks! We only need at most two items, though.

**Comment by CyberShadow** [comment]
So, how will this behave in case something fails?

File: dblog_feed_example.xml (hunk 3 above)
Comment (wilzbach) [reply dblog_feed_example.xml]: I thought if we already test this, then we should do it with an actual real world example.

**Comment by wilzbach** [comment]
> So, how will this behave in case something fails?

It will exit with a non-zero exit code before writing the file (hence `--output`) which means the entire deployment will fail.
This doesn't apply for HTML escaping errors though, but for everything else related to extracting the data from the XML.

File: tools/ddoc_xml_extractor.d (hunk 2 above)
Comment (wilzbach) [reply tools/ddoc_xml_extractor.d]: Yes you are right though I don't think we have HTML escaping out-of the box in Phobos, do we?
One could also make the point that dlang.org uses `<meta charset="utf-8">` and HTML escaping for non-ASCII characters isn't required.

Comment (CyberShadow) [reply tools/ddoc_xml_extractor.d]: `std.xml.encode` ?

Comment (CyberShadow) [reply tools/ddoc_xml_extractor.d]: >  One could also make the point that dlang.org uses <meta charset="utf-8"> and HTML escaping for non-ASCII characters isn't required.

Right, not unicode, just `&amp;` and `&gt;`

Comment (wilzbach) [reply tools/ddoc_xml_extractor.d]: Ah nice! Didn't know about that one. Done!

**Comment by CyberShadow** [comment]
> from Mon, 19 Mar 2018

Oof, I don't suppose we could easily use a more human-friendly date format there...

**Comment by wilzbach** [comment]
> Oof, I don't suppose we could easily use a more human-friendly date format there...

Well, std.datetime doesn't support this :/
Am I allowed to use http://code.dlang.org/packages/datefmt?

**Comment by jmdavis** [comment]
`std.datetime.systime.parseRFC822DateTime` would handle that particular monstrosity it if it had a time zone on it, but it doesn't, so it won't.

**Comment by wilzbach** [comment]
It does have a timezone attached (see the XML), but I think @CyberShadow was asking about a more human-friendly output like "March" instead of "Mar" or "Monday" instead of "Mon".

**Comment by CyberShadow** [comment]
> Am I allowed to use http://code.dlang.org/packages/datefmt?

Blah, that package has nothing on `ae.utils.time` :P

Formatting is actually much easier than parsing, could be done by hand. No library needed.

**Comment by wilzbach** [comment]
> Formatting is actually much easier than parsing, could be done by hand. No library needed.
> Yeah, formatting is generally

I agree and `std.format` handles most of this, but for things like `weekday.toPretty` or `month.toPretty` I just feel like I'm not the first guy needing to do this conversion.

Anyhow, I now added a hard-coded month "prettification", s.t. we can emulate the current output on dlang.org.

Anything else required from this PR?

**Comment by wilzbach** [comment]
(Updated)

**Comment by wilzbach** [comment]
Going to merge this now as the RegExp seemed to have failed again:

image

File: tools/ddoc_xml_extractor.d [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+auto toDdoc(const ref Article article, string suffix)
+{
+    import std.xml : encode;
+    string text;
+with(article)
+{
+    auto prettyMonth = ["January", "February", "March", "April", "May", "June", "July",
+    "August", "September", "October", "November", "December"][date.month-1];
```
Comment (CyberShadow) [inline tools/ddoc_xml_extractor.d]: Woah, what happened to the indentation here

Comment (wilzbach) [reply tools/ddoc_xml_extractor.d]: -> https://github.com/dlang/dlang.org/pull/2305

**Pull request merged**